sky/dag.py

```
@@ -2,74 +2,253 @@
+                          they depend on.
+        """
+        self.name = name
         self.tasks: List['task.Task'] = []
-        import networkx as nx  # pylint: disable=import-outside-toplevel
+        self._task_name_lookup: Dict[str, 'task.Task'] = {}
+        self.dependencies: Dict['task.Task', Set['task.Task']] = {}
 
         self.graph = nx.DiGraph()
-        self.name: Optional[str] = None
+
+        # Add tasks
+        if tasks:
+            for task in tasks:
+                self.add(task)
+
+        # Add dependencies
+        if dependencies:
+            for dependent, deps in dependencies.items():
+                self.set_dependencies(dependent, deps)
+
+    def _get_task(self, task_or_name: TaskOrName) -> 'task.Task':
+        """Get a task object from a task or its name.
+
+        Args:
+            task_or_name: Either a Task object or the name of a task.
+
+        Returns:
+            The Task object.
+
+        Raises:
+            ValueError: If the task name is not found in the DAG.
+        """
+        if not isinstance(task_or_name, str):
+            return task_or_name
+        name = task_or_name
+        if name not in self._task_name_lookup:
+            raise ValueError(f'Task {name} not found in DAG')
+        return self._task_name_lookup[name]
 
     def add(self, task: 'task.Task') -> None:
+        """Add a task to the DAG.
+
+        Args:
+            task: The Task object to add.
+
+        Raises:
+            ValueError: If the task already exists in the DAG or if its name
+            is already used.
+        """
+        if task in self.tasks:
+            raise ValueError(f'Task {task.name} already exists in the DAG.')
         self.graph.add_node(task)
         self.tasks.append(task)
+        if task.name is not None:
+            if task.name in self._task_name_lookup:
+                raise ValueError(
+                    f'Task name "{task.name}" is already used in the DAG.')
```

We should raise this error before we add the task to `self.graph`?

---

Quite.